### DISCUSSION TITLE
Is there a general explanation of the parameters and requirements for the functions in a notes backend?

### DISCUSSION THREAD

**Discussion by Hugo-Heagren** [emacs-citar/citar#703] [Q&A] [answered]
I want to write a notes backend for citar, using org-roam notes. I'm very close -- though I did quite a lot of the work a while ago and I think the api has changed a bit. I have looked at `citar-org-roam` briefly, but:
- it seems quite heavy for my usecase -- I don't need any of the extra features it implements
- it detects the relevant org roam node by including the id in the completiong string and finding it again with `split-string`. This seems brittle to me, and I would like to do arbitrary formatting on node title if possible.

I've bashed away at it for a few hours, but I'm getting quite frustrated at the system -- it's clearly very sophisticated but I can't tell what type of values my different functions are meant to have, or even really what they are meant to do. I want to know, for each :key in `citar-register-notes-source` (`:hasitems`,`:open`,`:items`,`:create`, `:transform`, `:annotate`):
- how many, and what type of arguments it will be passed
- what type of thing it must return (if this is constrained)
- how this related to the other functions (e.g. it doesn't matter what type of objects this function returns, so long as it can be handled as an arg by the function assigned at `:other-key`)

This would be really helpful!

**Comment by bdarcus** [comment, marked as answer]
I'm starting to fill out the docstrings in #704.

**Reply by Hugo-Heagren** [reply]
Thanks so much!
